nomad/fsm.go

```diff
@@ -625,6 +625,49 @@ func (n *nomadFSM) applyUpsertJob(msgType structs.MessageType, buf []byte, index
 		panic(fmt.Errorf("failed to decode request: %v", err))
 	}
 
+	// Preserve the existing task group counts and resources, if requested. Do
+	// this here to avoid race conditions when registering and scaling jobs.
+	if req.PreserveCounts || req.PreserveResources {
```

Comment:
@tgross I moved the logic, this seemed like the place it made the most sense but let me know if you had something else in mind!

Reply:
We try to keep logic out of `nomad/fsm.go` and push it down into the state store in `nomad/state` (which is also "in the FSM" but not in the `nomad/fsm.go` code... that's on me for being unclear when I suggested it). It makes for nicer testing and makes the FSM here more legible. There's an `UpsertJob` method on the state store that we call below. I would push this logic down into there.

Reply:
That's a good point. But there's definitely some precedent for passing the full request into the state store method. We do that in `UpsertPlanResults`, `UpsertScalingEvent`, `UpdateNodeStatus`, `UpdateDeploymentStatus` (and a handful of other deployment-related upserts), and probably some others. All of those are cases where we are trying to push behavior flags from the request down into the state store upsert.

One thing that makes that annoying is that `UpsertJob` has hundreds of callers because it's used a lot in tests. So we could provide a wrapper method that takes the request, does this work, and then defers to `UpsertJob`.